### PR TITLE
ARACHNE-2371 Add endpoint for bulk users registration

### PR DESCRIPTION
**Pull request by sergey-maletsky**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/portal/exception/EmailNotUniqueException.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.odysseusinc.arachne.portal.exception;
```
Comment (agackovka) [inline src/main/java/com/odysseusinc/arachne/portal/exception/EmailNotUniqueException.java]: copyright

File: src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java [hunk 2]
```diff
@@ -590,6 +598,21 @@ public void saveUsers(List<U> users, Set<Tenant> tenants, boolean emailConfirmat
         });
     }
 
+    private Map<String, String> getEmailValidationErrors(List<U> users) {
+
+        Map<String, String> emailValidationErrors = new HashMap<>();
+        IntStream.range(0, users.size())
+                .forEach(index -> {
+                    U user = users.get(index);
+                    U byEmail = getByUnverifiedEmailInAnyTenant(user.getEmail());
```
Comment (agackovka) [inline src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java]: i believe these users can be gathered in a single query

File: src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java [hunk 3]
```diff
@@ -590,6 +598,21 @@ public void saveUsers(List<U> users, Set<Tenant> tenants, boolean emailConfirmat
         });
     }
 
+    private Map<String, String> getEmailValidationErrors(List<U> users) {
+
+        Map<String, String> emailValidationErrors = new HashMap<>();
+        IntStream.range(0, users.size())
```
Comment (agackovka) [inline src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java]: you don't use any advantages of streams here... looks terrible)

File: src/main/java/com/odysseusinc/arachne/portal/exception/EmailNotUniqueException.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import edu.emory.mathcs.backport.java.util.Collections;
```
Comment (agackovka) [inline src/main/java/com/odysseusinc/arachne/portal/exception/EmailNotUniqueException.java]: what?

File: src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java [hunk 5]
```diff
@@ -574,6 +577,11 @@ public U updateInAnyTenant(U user) throws NotExistException {
     @Transactional(rollbackOn = Exception.class)
     public void saveUsers(List<U> users, Set<Tenant> tenants, boolean emailConfirmationRequired) {
 
+        Map<String, String> emailValidationErrors = getEmailValidationErrors(users);
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java]: Wouldn't it be easier to just rethrow NotUniqueException with proper key for UI (and exception handler)?

Comment (sergey-maletsky) [reply src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java]: we need to gather all email field errors that's why NotUniqueException will not be useful

**Pull request merged**